File: src/connector/Customer.java
package connector;

import java.util.ArrayList;
import java.util.HashMap;

import com.sap.conn.jco.JCoDestination;
import com.sap.conn.jco.JCoFunction;
import com.sap.conn.jco.JCoParameterList;
import com.sap.conn.jco.JCoStructure;
import com.sap.conn.jco.JCoTable;

public class Customer {
	
	public String number;
	public ArrayList<Invoice> invoices;
	
	private Connector connector;
	public Organization organization;
	
	public String
	firstName,
	lastName,
	city,
	district,
	postBox,
	postalCode,
	countryCode,
	cityCode,
	street,
	telephoneNumber,
	postBoxPostalCode;
	
	public Customer(String number, Connector connector)
	{
		this.number = number;
		this.connector = connector;
		this.organization = connector.organization;
		GetCustomerDetails();
		invoices = Invoices(number);
		
		// check retrieved data
		for (Invoice invoice : invoices)
		{
			System.out.println(invoice.toString());
			System.out.println("------------------------------");
		}
	}
	
	private ArrayList<Invoice> Invoices(String customerNumber)
    {
    	customerNumber = Utility.Pad(customerNumber,10,'0');
    	ArrayList<HashMap<String,String>> salesOrders = GetSalesOrders(customerNumber);
    	ArrayList<Invoice> result = new ArrayList<Invoice>();
    	for (HashMap<String,String> salesOrder : salesOrders)
    	{
    		result.add(new Invoice(this, salesOrder));
    	}
    	return result;
    }

	 public ArrayList<HashMap<String,String>> GetOrderDetails(String invoiceNumber)
     {	
    	try{
    		invoiceNumber = Utility.Pad(invoiceNumber, 10, '0');
    		JCoFunction f = connector.destination.getRepository().getFunction("BAPI_BILLINGDOC_GETLIST");
    		JCoParameterList inputs = f.getImportParameterList();
    		JCoStructure struc = inputs.getStructure("REFDOCRANGE");
    		struc.setValue("SIGN", "I");
    		struc.setValue("REF_DOC_LOW", invoiceNumber);
    		struc.setValue("REF_DOC_HIGH", invoiceNumber);
    		struc.setValue("OPTION", "EQ");
    		f.execute(connector.destination);
    		System.out.println(f.getExportParameterList().getStructure("RETURN").getString("TYPE"));
    		System.out.println(f.getExportParameterList().getStructure("RETURN").getString("MESSAGE"));
    		JCoTable table = f.getTableParameterList().getTable("BILLINGDOCUMENTDETAIL");
    		return Utility.Table(table);
    	} catch(Exception e){}
    	return null;
     }
	 
	 public void GetCustomerDetails()
     {	
    	try{
    		JCoFunction f = connector.destination.getRepository().getFunction("BAPI_CUSTOMER_GETDETAIL2");
    		f.getImportParameterList().setValue("CUSTOMERNO", number);
    		f.execute(connector.destination);
    		
    		JCoParameterList exportParameters = f.getExportParameterList();
    		JCoStructure info = exportParameters.getStructure("CUSTOMERADDRESS");
    		
    		System.out.println(exportParameters.getStructure("RETURN").getString("TYPE"));
    		System.out.println(exportParameters.getStructure("RETURN").getString("MESSAGE"));
    		
    		
    		firstName = info.getString("NAME");
    		lastName = info.getString("NAME_2");
    		city = info.getString("CITY");
    		district = info.getString("DISTRICT");
    		postBox = info.getString("PO_BOX");
    		postBoxPostalCode = info.getString("POBX_PCD");
    		postalCode = info.getString("POSTL_CODE");
    		countryCode = info.getString("COUNTYCODE"); 	
    		cityCode = info.getString("CITY_CODE");
    		street = info.getString("STREET");
    		telephoneNumber = info.getString("TELEPHONE");
    	} catch(Exception e){}
     }
	 
	    private ArrayList<HashMap<String,String>> GetSalesOrders(String customerNumber)
	    {
	    	
	    	try{
	    		JCoFunction f = connector.destination.getRepository().getFunction("BAPI_SALESORDER_GETLIST");
	    		JCoParameterList inputs = f.getImportParameterList();
	    		inputs.setValue("CUSTOMER_NUMBER", customerNumber); // 10 characters long - pad with leading zeros
	    		inputs.setValue("SALES_ORGANIZATION", "1000"); // 4 characters long - pad with leading zeros
	    		f.execute(connector.destination);
	    		System.out.println(f.getExportParameterList().getStructure("RETURN").getString("TYPE"));
	    		System.out.println(f.getExportParameterList().getStructure("RETURN").getString("MESSAGE"));
	    		JCoTable table = f.getTableParameterList().getTable("SALES_ORDERS");
	    		return Utility.Table(table);
	    		
	    	} catch(Exception e){}
	    	return null;
	    }
	    
	    public String toString()
	    {
	    	return
	    			Utility.EmptyOr(firstName) + 
	    			Utility.EmptyOr(lastName) + 
	    			Utility.EmptyOr(street) +
	    			Utility.EmptyOr(city);		
	    }
	
}

File: SAPTest/SAPTest/Customer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using SAP.Middleware.Connector;

namespace SAPTest
{
    class Customer
    {
        public List<Invoice> Invoices { get; set; }

        public Connector Connector { get; set; }
        public Organization Organization {get; set;}

        public String
        number,
        firstName,
        lastName,
        city,
        district,
        postBox,
        postalCode,
        countryCode,
        cityCode,
        street,
        telephoneNumber,
        postBoxPostalCode;

        public Customer(String num, Connector connector)
        {
            number = num.PadLeft(10, '0');
            Connector = connector;
            Organization = connector.Organization;
            GetCustomerDetails();
            Invoices = GetSalesOrders(number).Select(x => new Invoice(this, x)).ToList();
        }

        public void Check()
        {
            Console.WriteLine(ToString());
            Invoices.ForEach(i => Console.WriteLine($"{i.ToString()}\n---------------------------------"));
        }


        public List<Dictionary<String, String>> GetOrderDetails(String invoiceNumber)
        {
            try
            {
                invoiceNumber = invoiceNumber.PadLeft(10, '0');
                var f = Function("BAPI_BILLINGDOC_GETLIST");
                var struc = f.GetStructure("REFDOCRANGE");
                struc.SetValue("SIGN", "I");
                struc.SetValue("REF_DOC_LOW", invoiceNumber);
                struc.SetValue("REF_DOC_HIGH", invoiceNumber);
                struc.SetValue("OPTION", "EQ");
                f.Invoke(Connector.Destination);
                Result(f);
                var table = f.GetTable("BILLINGDOCUMENTDETAIL");
                return Util.Table(table);
            }
            catch (Exception _) {
                return new List<Dictionary<string, string>>();
            }
        }

        IRfcFunction Function(string f) => Connector.Destination.Repository.CreateFunction(f);

        static void Result(IRfcFunction f)
        {
            Console.WriteLine(f.GetStructure("RETURN").GetString("TYPE"));
            Console.WriteLine(f.GetStructure("RETURN").GetString("MESSAGE"));
        }

        public void GetCustomerDetails()
        {
            try
            {
                var f = Function("BAPI_CUSTOMER_GETDETAIL2");
                f.SetValue("CUSTOMERNO", number);
                f.Invoke(Connector.Destination);
                var info = f.GetStructure("CUSTOMERADDRESS");
                Result(f);
                firstName = info.GetString("NAME");
                lastName = info.GetString("NAME_2");
                city = info.GetString("CITY");
                district = info.GetString("DISTRICT");
                postBox = info.GetString("PO_BOX");
                postBoxPostalCode = info.GetString("POBX_PCD");
                postalCode = info.GetString("POSTL_CODE");
                countryCode = info.GetString("COUNTYCODE");
                cityCode = info.GetString("CITY_CODE");
                street = info.GetString("STREET");
                telephoneNumber = info.GetString("TELEPHONE");
            }
            catch (Exception _) {  }
        }

        private List<Dictionary<String, String>> GetSalesOrders(String customerNumber)
        {

            try
            {
                var f = Function("BAPI_SALESORDER_GETLIST");
                f.SetValue("CUSTOMER_NUMBER", customerNumber); 
                f.SetValue("SALES_ORGANIZATION", "1000");
                f.Invoke(Connector.Destination);
                Result(f);
                return Util.Table(f.GetTable("SALES_ORDERS"));

            }
            catch (Exception e)
            {
                return new List<Dictionary<string, string>>();
            }
        }

        public override String ToString()
        {
            return $"{firstName}\n{lastName}\n{street}\n{city}";
        }

    }
}
